File: src/main/java/com/home/controllers/Controller.java
package com.home.controllers;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.home.domain.DataRepository;
import com.home.dto.Aggregate;
import com.home.dto.Data;
import com.home.integration.RemoteDataClient;
import retrofit2.Retrofit;
import retrofit2.converter.jackson.JacksonConverterFactory;
import spark.Request;
import spark.Response;
import spark.Spark;

import javax.inject.Inject;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;

public class Controller {

    private final ObjectMapper objectMapper;
    private final ExecutorService executorService;

    @Inject
    private DataRepository dataRepository;

    private final RemoteDataClient remoteDataClient = new RemoteDataClient(new Retrofit.Builder()
            .baseUrl("http://localhost:4567/")
            .addConverterFactory(JacksonConverterFactory.create())
            .build());


    public Controller(ObjectMapper objectMapper, ExecutorService executorService) {
        this.objectMapper = objectMapper;
        this.executorService = executorService;
    }

    public Data getData(Request request, Response response) {
        response.type("application/json");

        Long id = Long.parseLong(request.params("id"));
        Data data = dataRepository.read(id);
        int status = data != null ? 200 : 404;
        response.status(status);
        return data;
    }

    public Data deleteData(Request request, Response response) {
        Long id = Long.parseLong(request.params("id"));

        Data old = dataRepository.delete(id);
        int status = old != null ? 200 : 404;
        response.status(status);
        return old;
    }

    public Data createData(Request request, Response response) throws IOException {
        String body = request.body();
        Data data = objectMapper.readValue(body, Data.class);

        dataRepository.create(data);
        response.status(201);
        return data;
    }

    public Data updateData(Request request, Response response) throws IOException {

        Long id = Long.parseLong(request.params("id"));
        String body = request.body();
        Data data = objectMapper.readValue(body, Data.class);

        if (!data.getId().equals(id)) {
            Spark.halt(400);
        }

        Data old = dataRepository.read(id);
        int status = old != null ? 201 : 200;

        dataRepository.update(data);
        response.status(status);
        return data;
    }

    public Aggregate getAggregate(Request request, Response response) throws ExecutionException, InterruptedException {

        Future<Data> data1Future = executorService.submit(() -> remoteDataClient.getRemoteData("1", 600));
        Data data2 = remoteDataClient.getRemoteData("2", 600);

        List<Data> results = Arrays.asList(data1Future.get(), data2);
        return new Aggregate(results);
    }

}

File: src/main/kotlin/com/home/controllers/Controller.kt
package com.home.controllers

import com.fasterxml.jackson.databind.ObjectMapper
import com.home.context
import com.home.domain.DataRepository
import com.home.dto.Aggregate
import com.home.dto.Data
import com.home.integration.DataBaseClient
import spark.Request
import spark.Response
import spark.Spark
import java.io.IOException

class Controller(val repository: DataRepository,
                 val objectMapper: ObjectMapper = context().objectMapper,
                 val dataBaseClient: DataBaseClient = context().dbClient) {

    private fun Request.getId(): Long {
        return this.params("id").toLong()
    }

    private fun Request.getData(): Data {
        val body = this.body()
        return objectMapper.readValue(body, Data::class.java)
    }

    private fun Response.setStatusBasedOnPresence(any: Any?) {
        if (any == null) {
            this.status(404)
        } else {
            this.status(200)
        }
    }

    fun getData(request: Request, response: Response): Data? {
        val data = repository.read(request.getId())
        response.setStatusBasedOnPresence(data)
        return data
    }

    fun deleteData(request: Request, response: Response): Data? {
        val data = repository.delete(request.getId())
        response.setStatusBasedOnPresence(data)
        return data
    }


    @Throws(IOException::class)
    fun createData(request: Request, response: Response): Data {
        val data = request.getData()
        repository.create(data)
        response.status(201)
        return data
    }

    @Throws(IOException::class)
    fun updateData(request: Request, response: Response): Data {
        val id = request.getId()
        val data = request.getData()

        if (id != data.id) {
            Spark.halt(400)
        }

        val old = repository.read(id)
        val status = if (old != null) 201 else 200

        repository.update(data)
        response.status(status)
        return data
    }

    @Throws(IOException::class)
    fun getAggregate(@Suppress("UNUSED_PARAMETER") request: Request, response: Response): Aggregate {

        dataBaseClient.proceedCreateUser()
        dataBaseClient.proceedGetUsers()

        val data1 = repository.read(1)
        val data2 = repository.read(2)

        response.status(200)
        return Aggregate(setOf(data1, data2))
    }

}
